internal/olm/operator/internal/registry_pod.go

```diff
@@ -0,0 +1,276 @@
+	pod, err := rp.build()
+	if err != nil {
+		return nil, fmt.Errorf("error in building registry pod: %v", err)
+	}
+
+	// Check if Pod already exists
+	tmpPod, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Create pod in kubernetes cluster using the above pod definition and clientset
+			pod, err = rp.Kubeclient.CoreV1().Pods(pod.Namespace).Create(ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("error creating registry pod: %v", err)
+			}
+		} else {
+			return nil, fmt.Errorf("error getting existing registry pod: %v", err)
+		}
+	} else {
+		return tmpPod, nil
+	}
+
+	return pod, nil
+}
+
+func (rp *RegistryPod) verifyPodCreation(ctx context.Context, pod *corev1.Pod) error {
+	// Upon creation of new pod, poll and verify that pod status is running
+	podCheck := wait.ConditionFunc(func() (done bool, err error) {
+		p, err := rp.Kubeclient.CoreV1().Pods(pod.Namespace).Get(ctx, pod.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, fmt.Errorf("error getting pod %s: %w", pod.Name, err)
+		}
+		return p.Status.Phase == corev1.PodRunning, nil
+	})
+
+	err := wait.PollImmediateUntil(time.Duration(1*time.Second), podCheck, ctx.Done())
+	if err != nil {
+		return fmt.Errorf("error waiting for registry pod %s to run: %v", pod.Name, err)
+	}
+
+	_, err = rp.fetchPodLogs(ctx, pod)
+	if err != nil {
+		return fmt.Errorf("error in fetching pod logs: %v", err)
+	}
+
+	return err
+}
+
+func (rp *RegistryPod) init() error {
+	rp.setDefaults()
+	err := rp.validate()
+	if err != nil {
+		return fmt.Errorf("error in validating registry pod: %v", err)
+	}
+	return nil
+}
+
+func (rp *RegistryPod) setDefaults() {
+	// set the grpcPort to default 50051
+	rp.grpcPort = defaultGRPCPort
+
+	if len(rp.IndexImage) == 0 {
+		rp.IndexImage = defaultIndexImage
+	}
+
+	if len(rp.BundleAddMode) == 0 {
+		if rp.IndexImage == defaultIndexImage {
+			rp.BundleAddMode = "semver"
+		} else {
+			rp.BundleAddMode = "replaces"
+		}
+	}
+}
+
+func (rp *RegistryPod) validate() error {
+	if len(rp.BundleImage) == 0 {
+		return errors.New("bundle image is required")
+	}
+	if len(rp.DBPath) == 0 {
+		return errors.New("registry database path is required")
+	}
+
+	if len(rp.Namespace) == 0 {
+		return errors.New("pod namespace is required")
+	}
+
+	return nil
+}
+
+// getPodName will return a string constructed from the bundle Image name
+func getPodName(bundleName string) (string, error) {
+	var podName string
+	if strings.Contains(bundleName, "/") {
+		split := strings.Split(bundleName, "/")
+		lastSegment := strings.Split(split[len(split)-1:][0], ":")
+		// get the last but one element from last segment excluding
+		// the version number and append index
+		podName = lastSegment[len(lastSegment)-2] + "-index"
+	} else {
+		return "", fmt.Errorf("invalid bundle image name: %s", bundleName)
+	}
+	return podName, nil
```

Comment:
I don't think this logic is foolproof. Consider the case where the bundle image name is: `quay.io/myproj/myimage@sha256:5891b5b522d5df086d0ff0b110fbd9d21bb4fc7163af34d08286a2e846f6be03`

It might be better to use a library to parse the image reference. E.g. [`sigs.k8s.io/cluster-api/util/container.ImageFromString()`](https://pkg.go.dev/sigs.k8s.io/cluster-api/util/container?tab=doc#ImageFromString)